### PR TITLE
Improve LineReader with custom delimiters

### PR DESCRIPTION
**Pull request by userqq**
Please don't be angry for [$line, $this->buffer] thing, I will change it if you let me know.
Just tried to simplify code a bit.

### CODE REVIEW

File: lib/LineReader.php [hunk 1]
```diff
@@ -7,15 +7,21 @@
...
+    private $ending;
+
     /** @var string */
     private $buffer = "";
 
     /** @var InputStream */
     private $source;
 
-    public function __construct(InputStream $inputStream)
+    public function __construct(InputStream $inputStream, string $ending = self::DEFAULT_ENDING)
```
Comment (kelunik) [inline lib/LineReader.php:21]: I think we should default to `null` here and allow passing `"\n"` explicitly to disable the trimming.

File: lib/LineReader.php [hunk 2]
```diff
@@ -24,19 +30,17 @@ public function __construct(InputStream $inputStream)
...
             while (null !== $chunk = yield $this->source->read()) {
                 $this->buffer .= $chunk;
 
-                if (($pos = \strpos($this->buffer, "\n")) !== false) {
-                    $line = \substr($this->buffer, 0, $pos);
-                    $this->buffer = \substr($this->buffer, $pos + 1);
-                    return \rtrim($line, "\r");
+                if (false !== \strpos($this->buffer, $this->ending)) {
+                    [$line, $this->buffer] = explode($this->ending, $this->buffer, 2);
+                    return $this->ending === self::DEFAULT_ENDING ? \rtrim($line, "\r") : $line;
```
Comment (kelunik) [inline lib/LineReader.php:43]: We can store a property for this if `$ending` is `null` in the constructor.

File: lib/LineReader.php [hunk 3]
```diff
@@ -7,15 +7,21 @@
 
 final class LineReader
 {
+    const DEFAULT_ENDING = "\n";
+
+    /** @var string */
+    private $ending;
```
Comment (kelunik) [inline lib/LineReader.php:13]: I'd rename this property to `$delimiter`.

**Comment by userqq** [comment]
@kelunik 
Please see updated.

File: lib/LineReader.php [hunk 4]
```diff
@@ -7,21 +7,23 @@
...
 final class LineReader
 {
-    const DEFAULT_ENDING = "\n";
-
     /** @var string */
-    private $ending;
+    private $delimiter;
+
+    /** @var bool */
+    protected $lineMode;
```
Comment (kelunik) [inline lib/LineReader.php:14]:
```suggestion
    private $lineMode;
```

File: lib/LineReader.php [hunk 5]
```diff
@@ -7,15 +7,23 @@
...
     private $buffer = "";
 
     /** @var InputStream */
     private $source;
 
-    public function __construct(InputStream $inputStream)
+    public function __construct(InputStream $inputStream, string $delimiter = null)
     {
-        $this->source = $inputStream;
+        $this->source    = $inputStream;
```
Comment (kelunik) [inline lib/LineReader.php:24]:
```suggestion
        $this->source = $inputStream;
```

File: lib/LineReader.php [hunk 6]
```diff
@@ -7,15 +7,23 @@
...
+        $this->delimiter = $delimiter === null ? "\n" : $delimiter;
+        $this->lineMode  = $delimiter === null;
```
Comment (kelunik) [inline lib/LineReader.php:26]:
```suggestion
        $this->lineMode = $delimiter === null;
```

File: lib/LineReader.php [hunk 7]
```diff
@@ -24,19 +32,17 @@ public function __construct(InputStream $inputStream)
     public function readLine(): Promise
     {
         return call(function () {
-            if (($pos = \strpos($this->buffer, "\n")) !== false) {
-                $line = \substr($this->buffer, 0, $pos);
-                $this->buffer = \substr($this->buffer, $pos + 1);
-                return \rtrim($line, "\r");
+            if (false !== strpos($this->buffer, $this->delimiter)) {
+                [$line, $this->buffer] = explode($this->delimiter, $this->buffer, 2);
```
Comment (kelunik) [inline lib/LineReader.php:36]:
```suggestion
            if (false !== \strpos($this->buffer, $this->delimiter)) {
                [$line, $this->buffer] = \explode($this->delimiter, $this->buffer, 2);
```

File: lib/LineReader.php [hunk 8]
```diff
@@ -24,19 +32,17 @@ public function __construct(InputStream $inputStream)
...
 
             while (null !== $chunk = yield $this->source->read()) {
                 $this->buffer .= $chunk;
 
-                if (($pos = \strpos($this->buffer, "\n")) !== false) {
-                    $line = \substr($this->buffer, 0, $pos);
-                    $this->buffer = \substr($this->buffer, $pos + 1);
-                    return \rtrim($line, "\r");
+                if (false !== \strpos($this->buffer, $this->delimiter)) {
+                    [$line, $this->buffer] = explode($this->delimiter, $this->buffer, 2);
```
Comment (kelunik) [inline lib/LineReader.php:44]:
```suggestion
                    [$line, $this->buffer] = \explode($this->delimiter, $this->buffer, 2);
```

**Comment by userqq** [comment]
@kelunik Thank you for review!

File: lib/LineReader.php [hunk 9]
```diff
@@ -24,19 +32,17 @@ public function __construct(InputStream $inputStream)
     public function readLine(): Promise
     {
         return call(function () {
-            if (($pos = \strpos($this->buffer, "\n")) !== false) {
-                $line = \substr($this->buffer, 0, $pos);
-                $this->buffer = \substr($this->buffer, $pos + 1);
-                return \rtrim($line, "\r");
+            if (false !== \strpos($this->buffer, $this->delimiter)) {
+                [$line, $this->buffer] = \explode($this->delimiter, $this->buffer, 2);
```
Comment (kelunik) [inline lib/LineReader.php:36]:
```suggestion
                list($line, $this->buffer) = \explode($this->delimiter, $this->buffer, 2);
```

File: lib/LineReader.php [hunk 10]
```diff
@@ -24,19 +32,17 @@ public function __construct(InputStream $inputStream)
...
 
             while (null !== $chunk = yield $this->source->read()) {
                 $this->buffer .= $chunk;
 
-                if (($pos = \strpos($this->buffer, "\n")) !== false) {
-                    $line = \substr($this->buffer, 0, $pos);
-                    $this->buffer = \substr($this->buffer, $pos + 1);
-                    return \rtrim($line, "\r");
+                if (false !== \strpos($this->buffer, $this->delimiter)) {
+                    [$line, $this->buffer] = \explode($this->delimiter, $this->buffer, 2);
```
Comment (kelunik) [inline lib/LineReader.php:44]:
```suggestion
                    list($line, $this->buffer) = \explode($this->delimiter, $this->buffer, 2);
```

**Comment by kelunik** [comment]
@userqq @hlib-kuznetsov Seems like committer and PR author account don't match, I hope that's fine.

**Comment by hlib-kuznetsov** [comment]
@kelunik Oh, sorry for confusing, it's both my accounts, i've created PR from this account by mistake.

**Pull request merged**